Content.Server/Fluids/EntitySystems/MoppingSystem.cs

```diff
@@ -0,0 +1,363 @@
+                {
+                    msg = "mopping-system-target-container-too-small";
+                    user.PopupMessage(user, Loc.GetString(msg, ("target", target))); // play message now because we are aborting.
+                    return;
+                }
+                else
+                {
+                    // Determine transferAmount
+                    if (_tagSystem.HasTag(used, "Mop") // if the tool used is a literal mop (and not a sponge, rag, etc.)
+                        && !_tagSystem.HasTag(target, "Wringer")) // and if the target does not have a wringer for properly drying the mop
+                    {
+                        delay = 5.0f; // Should take much longer if you don't have a wringer
+
+                        if ((currentVolume / (currentVolume + availableVolume) ) > 0.25) // mop is more than one-quarter full
+                        {
+                            transferAmount = FixedPoint2.Min(refillableSolution.AvailableVolume, currentVolume * 0.6); // squeeze up to 60% of the solution from the mop.
+                            msg = "mopping-system-hand-squeeze-little-wet";
+
+                            if ((currentVolume / (currentVolume + availableVolume) ) > 0.5) // if the mop is more than half full
+                                msg = "mopping-system-hand-squeeze-still-wet"; // overwrites the above
+
+                        }
+                        else // mop is less than one-quarter full
+                        {
+                            transferAmount = FixedPoint2.Min(refillableSolution.AvailableVolume, currentVolume); // squeeze remainder of solution from the mop.
+                            msg = "mopping-system-hand-squeeze-dry";
+                        }
+
+                    }
+                    else
+                    {
+                        transferAmount = FixedPoint2.Min(refillableSolution.AvailableVolume, currentVolume); //Transfer all liquid from the tool to the container, but only if it will fit.
+                        msg = "mopping-system-refillable-success";
+                        delay = 1.0f;
+                    }
+
+                    donor = used; // the mop/tool Uid
+                    donorSolutionName = "absorbed"; // by definition on AbsorbentComponent
+
+                    acceptor  = target; // the refillable container's Uid
+                    acceptorSolutionName = refillable.Solution;
+
+                    // Set delay/popup/sound if nondefault. Popup and sound will only play on a successful doAfter.
+
+                    sfx = component.TransferSound;
+                }
+            }
+            else return; // if the target did not have a Refillable Component
+        }
+        else if (currentVolume <= 0) // tool is dry
+        {
+            if (TryComp<DrainableSolutionComponent>(target, out var drainable))
+            {
```

Comment:
probably combine these into one statement, same with the one above this for refillable

Reply:
the control flow for this whole function is reaaaally dirty and indented way too much in some places, probably needs a cleanup overall. makes it a little hard to understand

Reply:
That's fair, I'll see what I can do. Could probably start by combining the checks at 237 and 239 with an `&&`.

Reply:
Oh actually, how about I separate the big Puddle-Refillable-Drainable block into three functions (one to accept each type of component)? That might make it easier to follow.

Reply:
Scratch my last comment, I'll just do as Acruid now suggests. I did however combine this pair of checks into one statement, and same for refillable.